packages/exchange/src/pairs/XDaiBridge.ts

```diff
@@ -1,39 +1,126 @@
-import Pair, { ExchangeParams, ValueTypes } from './Pair';
+import Bridge from "./Bridge";
+import { EventData } from "web3-eth-contract";
+import { AbiItem } from "web3-utils";
 
-const toXdaiBridge = '0x4aa42145Aa6Ebf72e164C9bBC74fbD3788045016';
-const toDaiBridge = '0x7301cfa0e1756b71869e93d4e4dca5c7d0eb0aa6';
-
-export default class XDaiBridge extends Pair {
-  constructor() {
-    super({ assetA: 'xdai', assetB: 'dai' });
+const bridgeAAbi: AbiItem[] = [
+  {
+    "anonymous": false,
+    "inputs": [
+      {
+        "indexed": false,
+        "name": "recipient",
+        "type": "address"
+      },
+      {
+        "indexed": false,
+        "name": "value",
+        "type": "uint256"
+      },
+      {
+        "indexed": false,
+        "name": "transactionHash",
+        "type": "bytes32"
+      }
+    ],
+    "name": "AffirmationCompleted",
+    "type": "event"
   }
+]
 
-  exchangeAtoB({ account, value, ether }: ExchangeParams) {
-    const _value = this._getValue({ value, ether });
-    const xdai = this.getExchange().getAsset('xdai');
-    return xdai.send({
-      from: account,
-      value: _value,
-      to: toDaiBridge,
-    });
+const bridgeBAbi: AbiItem[] = [
+  {
+    "anonymous": false,
+    "inputs": [
+      {
+        "indexed": false,
+        "name": "recipient",
+        "type": "address"
+      },
+      {
+        "indexed": false,
+        "name": "value",
+        "type": "uint256"
+      },
+      {
+        "indexed": false,
+        "name": "transactionHash",
+        "type": "bytes32"
+      }
+    ],
+    "name": "RelayedMessage",
+    "type": "event"
   }
+]
 
-  exchangeBtoA({ account, value, ether }: ExchangeParams) {
-    const _value = this._getValue({ value, ether });
+const wait = (time: number) => new Promise(resolve => setTimeout(resolve, time));
 
-    const dai = this.getExchange().getAsset(this.assetB);
-    return dai.send({
-      from: account,
-      value: _value,
-      to: toXdaiBridge,
+export default class XDaiBridge extends Bridge {
+  constructor() {
+    super({
+      assetA: 'xdai',
+      assetABridge: '0x7301cfa0e1756b71869e93d4e4dca5c7d0eb0aa6',
+      assetB: 'dai',
+      assetBBridge: '0x4aa42145Aa6Ebf72e164C9bBC74fbD3788045016'
     });
   }
 
-  async estimateAtoB(value: ValueTypes) {
-    return this._getValue(value);
+  async detectExchangeBToAFinished(account: string, value: string, sendResult: any) {
+    const asset = this.getExchange().getAsset(this.assetA);
+    const web3 = asset.getWeb3();
+    const contract = new web3.eth.Contract(bridgeAAbi, this.assetABridge);
+    let fromBlock = await web3.eth.getBlockNumber()
+
+    while (true) {
```

Comment:
Is it suitable to wait forever? Does it makes sense to implement timeout? Or probably a user will have another approaches to interrupt the wait loop (e.g. by returning to the main page of the wallet), won't he/she?.

Reply:
Thanks, I think you are right. I added a timeout https://github.com/burner-wallet/burner-wallet-2/pull/21/commits/2948626aecd0dd32be7175afa09eca8d2b02abce